File: 77.combinations.py
#
# @lc app=leetcode id=77 lang=python
#
# [77] Combinations
#
from typing import *
import copy
# @lc code=start
class Solution(object):
    def __init__(self):
        self.res = []
        self.out = []
    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """
        self.helper(n,k,1)
        return self.res 

    def helper(self,n,k,level):
        if len(self.out) == k:
            self.res.append(copy.copy(self.out))
            return
        for i in range(level,n+1):
            self.out.append(i)
            self.helper(n,k,i+1)
            self.out.pop()
            
# @lc code=end

if __name__ =="__main__":
    print(Solution().combine(n=4,k=2))

File: golang/77.combinations.go
/*
 * @lc app=leetcode.cn id=77 lang=golang
 *
 * [77] Combinations
 */
package main

// @lc code=start
func combine(n int, k int) [][]int {
	nums := []int{}
	for i := 1; i <= n; i++ {
		nums = append(nums, i)
	}
	cur := []int{}
	ret := [][]int{}
	collect := func(item []int) {
		tmp := make([]int, len(item))
		copy(tmp, item)
		ret = append(ret, tmp)

	}
	combineHelper(nums, cur, 0, k, collect)
	return ret
}

// 排列限制了顺序就是组合
// func combineHelper(nums, cur []int, curIndex, k int, collect func([]int)) {
// 	if len(cur) == k {
// 		collect(cur)
// 		return
// 	}
// 	for i := curIndex; i < len(nums); i++ {
// 		cur = append(cur, nums[i])
// 		combineHelper(nums, cur, curIndex+1, k, collect)
// 		cur = cur[:len(cur)-1]
// 	}
// }
func combineHelper(nums, cur []int, curIndex, k int, collect func([]int)) {
	if len(cur) == k {
		collect(cur)
		return
	}
	for i := curIndex; i < len(nums); i++ {
		cur = append(cur, nums[i])
		combineHelper(nums, cur, i+1, k, collect)
		cur = cur[:len(cur)-1]
	}
}

// @lc code=end
